sessiongate/pkg/controller/controlplane/controller.go

```diff
@@ -0,0 +1,640 @@
+// Copyright 2025 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"golang.org/x/time/rate"
+	securityv1beta1 "istio.io/client-go/pkg/apis/security/v1beta1"
+	istioclientset "istio.io/client-go/pkg/clientset/versioned/typed/security/v1beta1"
+	istioinformers "istio.io/client-go/pkg/informers/externalversions/security/v1beta1"
+	istiolisters "istio.io/client-go/pkg/listers/security/v1beta1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	metav1apply "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	azcorearm "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+
+	sessiongatev1alpha1 "github.com/Azure/ARO-HCP/sessiongate/pkg/apis/sessiongate/v1alpha1"
+	"github.com/Azure/ARO-HCP/sessiongate/pkg/controller"
+	sessiongateapply "github.com/Azure/ARO-HCP/sessiongate/pkg/generated/applyconfiguration/sessiongate/v1alpha1"
+	clientset "github.com/Azure/ARO-HCP/sessiongate/pkg/generated/clientset/versioned"
+	sessiongateschema "github.com/Azure/ARO-HCP/sessiongate/pkg/generated/clientset/versioned/scheme"
+	informers "github.com/Azure/ARO-HCP/sessiongate/pkg/generated/informers/externalversions/sessiongate/v1alpha1"
+	listers "github.com/Azure/ARO-HCP/sessiongate/pkg/generated/listers/sessiongate/v1alpha1"
+	"github.com/Azure/ARO-HCP/sessiongate/pkg/mc"
+)
+
+// control plane controller implementation for Session resources.
+// it runs with leader election and handles Session reconciliation into
+// - istio AuthorizationPolicies
+// - session credentials secrets
```

Comment:
look at the validating admission policies used in the secret sync controller or the acr pull binding controller for how to secure this component from having root on the cluster